Source/RageCommunity.Library/Graphics/Checkpoint.cs

```diff
@@ -0,0 +1,223 @@
+                          Single radius = 1f,
+                          Single height = 1f,
+                          Int32 checkpointType = 47,
+                          Int32 reserved = 0,
+                          Boolean setZtoGround = true) : this(position, position, color, radius, height, checkpointType, reserved, setZtoGround)
+        {
+        }
+
+        public Checkpoint(Vector3 position,
+                          Vector3 nextPosition,
+                          Color color,
+                          Single radius,
+                          Single height,
+                          Int32 checkpointType = 0,
+                          Int32 reserved = 0,
+                          Boolean setZtoGround = true)
+        {
+            CheckpointType = checkpointType;
+            this.position = position;
+            this.nextPosition = nextPosition;
+            Radius = radius;
+            this.color = color;
+            Reserved = reserved;
+            this.setToGround = setZtoGround;
+            this.height = height;
+
+            CreateCheckpoint();
+        }
+
+        /// <summary>
+        /// The checkpoint types represents how the checkpoints looks in the game. 
+        /// </summary>
+        /// <remarks>
+        /// Checkpoint type 47 for example is a simple cylinder with no arrow or number.
+        /// Visit the Grand Theft Auto V Native DB for a list of all checkpoint types. 
+        /// </remarks>
+        public Int32 CheckpointType { get; }
+
+        public Color Color
+        {
+            get => this.color;
+            set
+            {
+                this.color = value;
+                UpdateColor(value);
+            }
+        }
+
+        /// <summary>
+        /// The ID of this checkpoint.
+        /// </summary>
+        public Int32 Handle { get; private set; }
+
+        public Single Height
+        {
+            get => this.height;
+            set
+            {
+                this.height = value;
+                SetHeight(this.height, this.height, Radius);
+            }
+        }
+
+        public Boolean IsValid { get; private set; }
+
+        /// <summary>
+        /// The position of the checkpoint after this one. 
+        /// </summary>
+        public Vector3 NextPosition
+        {
+            get => this.nextPosition;
+            set
+            {
+                this.nextPosition = value;
+                ReCreateCheckpoint();
+            }
+        }
+
+        public Vector3 Position
+        {
+            get => this.position;
+            set
+            {
+                this.position = value;
+                ReCreateCheckpoint();
+            }
+        }
+
+        public Single Radius { get; }
+
+        /// <summary>
+        /// Some <see cref="CheckpointType"/>s can display an icon or number. The number or icon can be set via this Reserved property. 
+        /// </summary>
+        public Int32 Reserved { get; }
```

Comment:
Should we be able to update the icon/number for a checkpoint with a setter?

Reply:
Yes, I added this to the implementation.